### PR TITLE
Move state.txt to /dev/shm to prevent boot up failure in case of a power loss

### LINKED ISSUE #6: Power loss or crash breaks tear_down/startup implementation
If the host have accidentally  a power loss or a system crash, the `ExecStop=/usr/local/sbin/microfw tear_down` will not be called. This leads to a startup error if you boot up your machine again. 
The reason for that is the `ExecStartPre=/usr/local/sbin/microfw tear_down` will fail due to the fact there is actually nothing to do for `tear_down` but the `state.txt` still exists.

### PR DESCRIPTION
**Pull request by weratyr**
Move state.txt into RAM to make microfw more robust against power loss or system crashes. 

This PR is stacked on top of #9 , so it depend on it

Fixes #6

### CODE REVIEW

File: etc/microfw.service [hunk 1]
```diff
@@ -4,8 +4,7 @@ After=network.target docker.service
 
 [Service]
 Type=oneshot
-ExecStartPre=/usr/local/sbin/microfw tear_down
-ExecStart=/var/lib/microfw/setup.sh
+ExecStart=/usr/local/sbin/microfw apply_bootup
 RemainAfterExit=true
 ExecStop=/usr/local/sbin/microfw tear_down
 StandardOutput=journal
```
Comment (Svedrin) [inline etc/microfw.service:10]: I just googled if `/dev/shm` is actually the right place to use for this, and got [sent to /run instead](https://wiki.debian.org/ReleaseGoals/%20RunDirectory#Overview). If we add a `RuntimeDirectory` here like so, systemd will create a `/run/microfw`  for us to use:

```suggestion
StandardOutput=journal
RuntimeDirectory=microfw
```

`/run` is a ramdisk, so it'll achieve the right thing.

File: features/teardown.sh [hunk 2]
```diff
@@ -34,6 +34,7 @@ function ip6tables () {
 
 RUNNING_IN_CI=true
 VAR_DIR="$TEMPDIR/var"
+SHM_DIR="$TEMPDIR/var"
```
Comment (Svedrin) [inline features/teardown.sh:37]:
```suggestion
RUN_DIR="$TEMPDIR/run"
```

File: src/generate_setup.py [hunk 3]
```diff
@@ -236,7 +237,7 @@ def printf(fmt, obj=None):
 
     def state(fmt, obj=None):
         """ generate a command to echo something into the state file """
-        printf('echo "%s" >> /var/lib/microfw/state.txt ' % fmt, obj)
+        printf('echo "%s" >> "%s"/state.txt ' % (fmt, SHM_DIR), obj)
```
Comment (Svedrin) [inline src/generate_setup.py:240]:
```suggestion
        printf('echo "%s" >> "%s"/state.txt ' % (fmt, RUN_DIR), obj)
```

File: src/microfw.sh [hunk 4]
```diff
@@ -8,6 +8,7 @@ if [ -z "${RUNNING_IN_CI:-}" ]; then
     COMMAND="invalid"
     ETC_DIR="/etc/microfw"
     VAR_DIR="/var/lib/microfw"
+    SHM_DIR="/dev/shm/microfw"
```
Comment (Svedrin) [inline src/microfw.sh:11]:
```suggestion
    RUN_DIR="/run/microfw"
```

File: src/microfw.sh [hunk 5]
```diff
@@ -54,7 +60,7 @@ if [ -z "${RUNNING_IN_CI:-}" ]; then
 fi
 
 function tear_down() {
-    if [ ! -e "$VAR_DIR/state.txt" ]; then
+    if [ ! -e "$SHM_DIR/state.txt" ]; then
```
Comment (Svedrin) [inline src/microfw.sh:63]:
```suggestion
    if [ ! -e "$RUN_DIR/state.txt" ]; then
```

File: src/microfw.sh [hunk 6]
```diff
@@ -65,8 +71,8 @@ function tear_down() {
     set +e
 
     # First let's find out if docker and mangle are present
-    HAVE_DOCKER="$(grep -q "^HAVE-DOCKER$" "$VAR_DIR/state.txt" && echo "true" || echo "false")"
-    HAVE_MANGLE="$(grep -q "^HAVE-MANGLE$" "$VAR_DIR/state.txt" && echo "true" || echo "false")"
+    HAVE_DOCKER="$(grep -q "^HAVE-DOCKER$" "$SHM_DIR/state.txt" && echo "true" || echo "false")"
+    HAVE_MANGLE="$(grep -q "^HAVE-MANGLE$" "$SHM_DIR/state.txt" && echo "true" || echo "false")"
```
Comment (Svedrin) [inline src/microfw.sh:75]:
```suggestion
    HAVE_DOCKER="$(grep -q "^HAVE-DOCKER$" "$RUN_DIR/state.txt" && echo "true" || echo "false")"
    HAVE_MANGLE="$(grep -q "^HAVE-MANGLE$" "$RUN_DIR/state.txt" && echo "true" || echo "false")"
```

File: src/microfw.sh [hunk 7]
```diff
@@ -102,7 +108,7 @@ function tear_down() {
     done
 
     # Flush and drop zone-specific chains
-    grep "^ZONE " "$VAR_DIR/state.txt" | while read command zone; do
+    grep "^ZONE " "$SHM_DIR/state.txt" | while read command zone; do
```
Comment (Svedrin) [inline src/microfw.sh:111]:
```suggestion
    grep "^ZONE " "$RUN_DIR/state.txt" | while read command zone; do
```

File: src/microfw.sh [hunk 8]
```diff
@@ -128,7 +134,7 @@ function tear_down() {
     ipset flush
     ipset destroy
 
-    rm "$VAR_DIR/state.txt"
+    rm "$SHM_DIR/state.txt"
```
Comment (Svedrin) [inline src/microfw.sh:137]:
```suggestion
    rm "$RUN_DIR/state.txt"
```

File: src/microfw.sh [hunk 9]
```diff
@@ -142,6 +148,7 @@ function generate_setup() {
 
 function compile() {
     mkdir -p $VAR_DIR
+    mkdir -p $SHM_DIR
```
Comment (Svedrin) [inline src/microfw.sh:151]:
```suggestion
    mkdir -p $RUN_DIR
```

**Review by Svedrin — CHANGES_REQUESTED** [review]

File: src/microfw.sh (hunk 9 above)
Comment (weratyr) [reply src/microfw.sh:151]: If systemd will create the directory `/run/microfw` we could get rid of the whole line.

**Comment by Svedrin** [comment]
we never got to finish this PR for some reason, but I like it, so I'm going to just merge as-is. yolo

**Pull request merged**